### PR TITLE
Do not disable gzip compression for nginx completely

### PR DESCRIPTION
**Pull request by benediktg**
see also issue #325

### CODE REVIEW

**Comment by benediktg** [comment]
Maybe somebody could remove irrelevant MIME types from the `gzip_types` directive?

**Comment by MorrisJobke** [comment]
I just tried this with the latest master and I could not find a request where this is valid, because we send an etag with every request somehow. So the gzip was never enabled. On the other side we enabled gzip for the static assets in the server itself.

I think this can be closed then.

**Comment by MorrisJobke** [comment]
> On the other side we enabled gzip for the static assets in the server itself.

See https://github.com/nextcloud/server/pull/4070

**Comment by benediktg** [comment]
Enabling gzip saves about 80 % data transfer and reduces the loading time by about 20 %:

screenshot_20170509_212601
screenshot_20170509_212619

([source](http://www.webpagetest.org/video/compare.php?tests=170509_GP_fcab759cf0d592d2f4f5f5387ffe34f2,170509_DF_3b5c769570b77dc17a8f7fc40d79724c))

**Comment by MorrisJobke** [comment]
With current master?

**Comment by MorrisJobke** [comment]
> With current master?

Because we do the GZIP compression in master already. ;)

**Comment by benediktg** [comment]
> With current master?

With 11.0.3

**Comment by MorrisJobke** [comment]
> With 11.0.3

There you have it ;)

**Comment by MorrisJobke** [comment]
> With 11.0.3

Just upgrade to the 12 beta 2 😉

**Comment by benediktg** [comment]
Here is the comparison with NC 12 beta 2:
screenshot_20170510_151323
screenshot_20170510_151340

([source](http://www.webpagetest.org/video/compare.php?tests=170510_E1_99090013a190a1a945c0bf0f1abcfde9,170510_87_82f46303d8cdb5f19ceccbdc858f5cb1))

**Comment by riegercloud** [comment]
I just performed some tests on Nextcloud 11.0.3 and as well on Nextcloud 12b2 ... Nextclouds performance significantly increased with NGINX 1.13 and ngx_cache_purge enabled. Thanks!!!

**Comment by MorrisJobke** [comment]
> Nextclouds performance significantly increased with NGINX 1.13 and ngx_cache_purge enabled. Thanks!!!

Could you show what you enabled so we can document it?

**Comment by riegercloud** [comment]
don't get me wrong but all i configured is shown at https://www.c-rieger.de/nextcloud-installation-guide/
regarding gzip i first tried your recommendations (gzip off) and had iterations from 

>gzip on;
gzip_vary on;
gzip_types application/javascript application/x-javascript text/javascript text/xml text/css;

to

> gzip on;
 gzip_min_length _**1024**_;
 gzip_proxied expired no-cache no-store private auth;
 gzip_types text/plain text/css text/xml text/javascript application/x-javascript application/json application/xml;
 gzip_disable msie6;

to finally yours:
> gzip on;
gzip_comp_level 4;
gzip_min_length 256;
gzip_proxied expired no-cache no-store private no_last_modified no_etag auth;
gzip_types application/atom+xml application/javascript application/json application/ld+json application/manifest+json application/rss+xml application/vnd.geo+json application/vnd.ms-fontobject application/x-font-ttf application/x-web-app-manifest+json application/xhtml+xml application/xml font/opentype image/bmp image/svg+xml image/x-icon text/cache-manifest text/css text/plain text/vcard text/vnd.rim.location.xloc text/vtt text/x-component text/x-cross-domain-policy;

**Comment by MorrisJobke** [comment]
> don't get me wrong but all i configured is shown at https://www.c-rieger.de/nextcloud-installation-guide/

Sorry :/ Didn't reminded that.

File: admin_manual/installation/nginx_nextcloud_9x.rst [hunk 1]
```diff
@@ -77,8 +77,12 @@ your nginx installation.
...
       fastcgi_buffers 64 4K;
   
-      # Disable gzip to avoid the removal of the ETag header
-      gzip off;
+      # Enable gzip but do not remove ETag headers
+      gzip on;
+      gzip_comp_level 4;
+      gzip_min_length 256;
+      gzip_proxied expired no-cache no-store private no_last_modified no_etag auth;
+      gzip_types application/atom+xml application/javascript application/json application/ld+json application/manifest+json application/rss+xml application/vnd.geo+json application/vnd.ms-fontobject application/x-font-ttf application/x-web-app-manifest+json application/xhtml+xml application/xml font/opentype image/bmp image/svg+xml image/x-icon text/cache-manifest text/css text/plain text/vcard text/vnd.rim.location.xloc text/vtt text/x-component text/x-cross-domain-policy;
```
Comment (MorrisJobke) [inline admin_manual/installation/nginx_nextcloud_9x.rst]: @josh4trunks Any opinions on this? Looks good from my side.

**Comment by MorrisJobke** [comment]
@benediktg Could you rebase this?

@LukasReschke Are you up to set this also up for apache? I guess it makes sense to gzip the static assets.

**Comment by josh4trunks** [comment]
I personally have my Nextcloud instance behind a Varnish cache server that does all the gzipping, and gzip is disabled for Nginx. There I gzip all `"(text|application)/"`

Is it posible to have **gzip_types** be something like `application/* text/*`? Otherwise this is probably fine, but I'll leave it to others to test.

**Comment by benediktg** [comment]
@morrisjobke Of course, I will do it soon.

**Comment by benediktg** [comment]
Do you think it would make sense to add
```nginx
gzip_vary on;
```
? (i.e. add the `Vary: Accept-Encoding` header)

**Comment by MorrisJobke** [comment]
> Do you think it would make sense to add

Yep - seems that it allows to make this compatible with clients that don't support gzip.

**Comment by MorrisJobke** [comment]
While working on this: Could you also add the same change to the subdir configuration below in the same file?

File: admin_manual/installation/nginx.rst [hunk 2]
```diff
@@ -90,8 +90,13 @@ webroot of your nginx installation. In this example it is
       client_max_body_size 512M;
       fastcgi_buffers 64 4K;
   
-      # Disable gzip to avoid the removal of the ETag header
-      gzip off;
+      # Enable gzip but do not remove ETag headers
+      gzip on;
+      gzip_vary on;
```
Comment (MorrisJobke) [inline admin_manual/installation/nginx.rst]: @josh4trunks Forgot to ask you: Any opinion on this? Works here fine and seems to make sense from my point of view.

**Comment by MorrisJobke** [comment]
I added it to the release notes: #f6497cf7

**Comment by josh4trunks** [comment]
**gzip_vary** looks good to me. As I said before, I do not personally use gzip with nginx, I let an upstream Varnish Cache proxy do that for me.

**Pull request merged**